src/optimizer/rules/unnesting_rules.cpp

```diff
@@ -105,4 +107,113 @@ void UnnestSingleJoinToInnerJoin::Transform(common::ManagedPointer<AbstractOptim
+///////////////////////////////////////////////////////////////////////////////
+DependentSingleJoinToInnerJoin::DependentSingleJoinToInnerJoin() {
+  type_ = RuleType::DEPENDENT_JOIN_GET_TO_INNER_JOIN;
+
+  match_pattern_ = new Pattern(OpType::LOGICALSINGLEJOIN);
+  match_pattern_->AddChild(new Pattern(OpType::LEAF));
+  auto right_child = new Pattern(OpType::LOGICALFILTER);
+  auto agg_child = new Pattern(OpType::LOGICALAGGREGATEANDGROUPBY);
+  agg_child->AddChild(new Pattern(OpType::LEAF));
+  right_child->AddChild(agg_child);
+  match_pattern_->AddChild(right_child);
+}
+
+RulePromise DependentSingleJoinToInnerJoin::Promise(GroupExpression *group_expr) const {
+  return RulePromise::LOGICAL_PROMISE;
+}
+
+bool DependentSingleJoinToInnerJoin::Check(common::ManagedPointer<AbstractOptimizerNode> plan,
+                                           OptimizationContext *context) const {
+  (void)context;
+  (void)plan;
+
+  UNUSED_ATTRIBUTE auto children = plan->GetChildren();
+  TERRIER_ASSERT(children.size() == 2, "SingleJoin should have 2 children");
+  return true;
+}
+
+void DependentSingleJoinToInnerJoin::Transform(common::ManagedPointer<AbstractOptimizerNode> input,
+                                               std::vector<std::unique_ptr<AbstractOptimizerNode>> *transformed,
+                                               UNUSED_ATTRIBUTE OptimizationContext *context) const {
+  UNUSED_ATTRIBUTE auto single_join = input->Contents()->GetContentsAs<LogicalSingleJoin>();
+  TERRIER_ASSERT(single_join->GetJoinPredicates().empty(), "SingleJoin should have no predicates");
+  // From LOGICALSINGLEJOIN -> LOGICALFILTER -> LOGICALAGGREGATEANDGROUPBY
+  //  to LOGICALFILTER -> LOGICALINNERJOIN -> LOGICALFILTER -> LOGICALAGGREGATEANDGROUPBY
+  auto &memo = context->GetOptimizerContext()->GetMemo();
+  auto filter_expr = input->GetChildren()[1];
+  auto agg_expr = filter_expr->GetChildren()[0];
+  auto agg_group_id = agg_expr->GetChildren()[0]->Contents()->GetContentsAs<LeafOperator>()->GetOriginGroup();
+  const auto &agg_group_aliases_set = memo.GetGroupByID(agg_group_id)->GetTableAliases();
+  auto &filter_predicates = filter_expr->Contents()->GetContentsAs<LogicalFilter>()->GetPredicates();
+
+  std::vector<AnnotatedExpression> top_predicates;
+  std::vector<AnnotatedExpression> down_predicates;
+  std::vector<common::ManagedPointer<parser::AbstractExpression>> new_groupby_cols;
+
+  // loop over all predicates check each of them if they refer table not contained in agg
+  for (auto &predicate : filter_predicates) {
+    if (OptimizerUtil::IsSubset(agg_group_aliases_set, predicate.GetTableAliasSet())) {
+      down_predicates.emplace_back(predicate);
+    } else {
+      // Correlated predicate, already in the form of
+      // (outer_relation.a = (expr))
+      top_predicates.emplace_back(predicate);
+      auto root_expr = predicate.GetExpr();
+      if (root_expr->GetChild(0)->GetDepth() < root_expr->GetDepth()) {
```

Comment:
Could you explain this section of code. I'm not fully clear on why this might come up.

Reply:
I commented above this block. This part of code is basically doing  RewritePullFilterThroughAggregation.

Reply:
I meant this if statement, codecov states it is also not covered by tests.

Reply:
This if statement still isn't covered and I don't think its changed since last time I reviewed.